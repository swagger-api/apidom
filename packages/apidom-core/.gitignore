/dist
/es
/cjs
/types
/NOTICE
/swagger-api-apidom-core-*.tgz
