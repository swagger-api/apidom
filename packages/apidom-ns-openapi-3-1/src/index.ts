export {
  isRefElement,
  isLinkElement as isLinkPrimitiveElement,
  isMemberElement,
  isObjectElement,
  isArrayElement,
  isBooleanElement,
  isNullElement,
  isElement,
  isNumberElement,
  isStringElement,
} from '@swagger-api/apidom-core';

export { default as mediaTypes, OpenAPIMediaTypes } from './media-types';

// eslint-disable-next-line no-restricted-exports
export { default } from './namespace';

export { default as refractorPluginReplaceEmptyElement } from './refractor/plugins/replace-empty-element';
export { default as refractorPluginNormalizeParameters } from './refractor/plugins/normalize-parameters';
export { default as refractorPluginNormalizeSecurityRequirements } from './refractor/plugins/normalize-security-requirements';
export { default as refractorPluginNormalizeServers } from './refractor/plugins/normalize-servers';
export { default as refractorPluginNormalizeOperationIds } from './refractor/plugins/normalize-operation-ids';
export { default as refractorPluginNormalizeParameterExamples } from './refractor/plugins/normalize-parameter-examples';
export { default as refractorPluginNormalizeHeaderExamples } from './refractor/plugins/normalize-header-examples';
export { default as createToolbox } from './refractor/toolbox';
export { default as specificationObj } from './refractor/specification';

export type {
  default as CallbackVisitor,
  CallbackVisitorOptions,
} from './refractor/visitors/open-api-3-1/callback';
export type {
  default as ComponentsVisitor,
  ComponentsVisitorOptions,
} from './refractor/visitors/open-api-3-1/components';
export type {
  default as ComponentsPathItemsVisitor,
  PathItemsVisitorOptions as ComponentsPathItemsVisitorOptions,
} from './refractor/visitors/open-api-3-1/components/PathItemsVisitor';
export type {
  default as ComponentsSchemasVisitor,
  SchemasVisitorOptions as ComponentsSchemasVisitorOptions,
} from './refractor/visitors/open-api-3-1/components/SchemasVisitor';
export type {
  default as ContactVisitor,
  ContactVisitorOptions,
} from './refractor/visitors/open-api-3-1/contact';
export type {
  default as DiscriminatorVisitor,
  DiscriminatorVisitorOptions,
} from './refractor/visitors/open-api-3-1/distriminator';
export type {
  default as EncodingVisitor,
  EncodingVisitorOptions,
} from './refractor/visitors/open-api-3-1/encoding';
export type {
  default as ExampleVisitor,
  ExampleVisitorOptions,
} from './refractor/visitors/open-api-3-1/example';
export type {
  default as ExternalDocumentationVisitor,
  ExternalDocumentationVisitorOptions,
} from './refractor/visitors/open-api-3-1/external-documentation';
export type {
  default as HeaderVisitor,
  HeaderVisitorOptions,
} from './refractor/visitors/open-api-3-1/header';
export type {
  default as InfoVisitor,
  InfoVisitorOptions,
} from './refractor/visitors/open-api-3-1/info';
export type {
  default as InfoSummaryVisitor,
  SummaryVisitorOption as InfoSummaryVisitorOption,
} from './refractor/visitors/open-api-3-1/info/SummaryVisitor';
export type {
  default as LicenseIdentifierVisitor,
  IdentifierVisitorOption as LicenseIdentifierVisitorOption,
} from './refractor/visitors/open-api-3-1/license/IdentifierVisitor';
export type {
  default as LicenseVisitor,
  LicenseVisitorOptions,
} from './refractor/visitors/open-api-3-1/license';
export type {
  default as LinkVisitor,
  LinkVisitorOptions,
} from './refractor/visitors/open-api-3-1/link';
export type {
  default as MediaTypeVisitor,
  MediaTypeVisitorOptions,
} from './refractor/visitors/open-api-3-1/media-type';
export type {
  default as OAuthFlowVisitor,
  OAuthFlowVisitorOptions,
} from './refractor/visitors/open-api-3-1/oauth-flow';
export type {
  default as OAuthFlowsVisitor,
  OAuthFlowsVisitorOptions,
} from './refractor/visitors/open-api-3-1/oauth-flows';
export type {
  default as OperationVisitor,
  OperationVisitorOptions,
} from './refractor/visitors/open-api-3-1/operation';
export type {
  default as ParameterVisitor,
  ParameterVisitorOptions,
} from './refractor/visitors/open-api-3-1/parameter';
export type {
  default as PathItemVisitor,
  PathItemVisitorOptions,
} from './refractor/visitors/open-api-3-1/path-item';
export type {
  default as PathsVisitor,
  PathsVisitorOptions,
} from './refractor/visitors/open-api-3-1/paths';
export type {
  default as ReferenceVisitor,
  ReferenceVisitorOptions,
} from './refractor/visitors/open-api-3-1/reference';
export type {
  default as RequestBodyVisitor,
  RequestBodyVisitorOptions,
} from './refractor/visitors/open-api-3-1/request-body';
export type {
  default as ResponseVisitor,
  ResponseVisitorOptions,
} from './refractor/visitors/open-api-3-1/response';
export type {
  default as ResponsesVisitor,
  ResponsesVisitorOptions,
} from './refractor/visitors/open-api-3-1/responses';
export type {
  default as Schema$defsVisitor,
  $defsVisitorOptions as Schema$defsVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema/$defsVisitor';
export type {
  default as Schema$refVisitor,
  $refVisitorOptions as Schema$refVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema/$refVisitor';
export type {
  default as Schema$vocabularyVisitor,
  $vocabularyVisitorOptions as Schema$vocabularyVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema/$vocabularyVisitor';
export type {
  default as SchemaAllOfVisitor,
  AllOfVisitorOptions as SchemaAllOfVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema/AllOfVisitor';
export type {
  default as SchemaAnyOfVisitor,
  AnyOfVisitorOptions as SchemaAnyOfVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema/AnyOfVisitor';
export type {
  default as SchemaDependentRequiredVisitor,
  DependentRequiredVisitorOptions as SchemaDependentRequiredVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema/DependentRequiredVisitor';
export type {
  default as SchemaDependentSchemasVisitor,
  DependentSchemasVisitorOptions as SchemaDependentSchemasVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema/DependentSchemasVisitor';
export type {
  default as SchemaEnumVisitor,
  EnumVisitorOptions as SchemaEnumVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema/EnumVisitor';
export type {
  default as SchemaExamplesVisitor,
  ExamplesVisitorOptions as SchemaExamplesVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema/ExamplesVisitor';
export type {
  default as SchemaVisitor,
  SchemaVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema';
export type {
  default as SchemaOneOfVisitor,
  OneOfVisitorOptions as SchemaOneOfVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema/OneOfVisitor';
export type {
  default as SchemaPatternPropertiesVisitor,
  PatternPropertiesVisitorOptions as SchemaPatternPropertiesVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema/PatternPropertiesVisitor';
export type {
  default as SchemaPrefixItemsVisitor,
  PrefixItemsVisitorOptions as SchemaPrefixItemsVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema/PrefixItemsVisitor';
export type {
  default as SchemaPropertiesVisitor,
  PropertiesVisitorOptions as SchemaPropertiesVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema/PropertiesVisitor';
export type {
  default as SchemaRequiredVisitor,
  RequiredVisitorOptions as SchemaRequiredVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema/RequiredVisitor';
export type {
  default as SchemaTypeVisitor,
  TypeVisitorOptions as SchemaTypeVisitorOptions,
} from './refractor/visitors/open-api-3-1/schema/TypeVisitor';
export type {
  default as SecurityRequirementVisitor,
  SecurityRequirementVisitorOptions,
} from './refractor/visitors/open-api-3-1/security-requirement';
export type {
  default as SecuritySchemeVisitor,
  SecuritySchemeVisitorOptions,
} from './refractor/visitors/open-api-3-1/security-scheme';
export type {
  default as ServerVisitor,
  ServerVisitorOptions,
} from './refractor/visitors/open-api-3-1/server';
export type {
  default as ServerVariableVisitor,
  ServerVariableVisitorOptions,
} from './refractor/visitors/open-api-3-1/server-variable';
export type {
  default as TagVisitor,
  TagVisitorOptions,
} from './refractor/visitors/open-api-3-1/tag';
export type {
  default as XmlVisitor,
  XmlVisitorOptions,
} from './refractor/visitors/open-api-3-1/xml';
export type {
  default as OpenApi3_1Visitor,
  OpenApi3_1VisitorOptions,
} from './refractor/visitors/open-api-3-1';
export type {
  default as JsonSchemaDialectVisitor,
  JsonSchemaDialectVisitorOptions,
} from './refractor/visitors/open-api-3-1/JsonSchemaDialectVisitor';
export type {
  default as WebhooksVisitor,
  WebhooksVisitorOptions,
} from './refractor/visitors/open-api-3-1/WebhooksVisitor';

export {
  isCallbackElement,
  isComponentsElement,
  isInfoElement,
  isJsonSchemaDialectElement,
  isLicenseElement,
  isOpenapiElement,
  isOpenApi3_1Element,
  isOperationElement,
  isParameterElement,
  isPathItemElement,
  isPathItemElementExternal,
  isReferenceElement,
  isReferenceElementExternal,
  isResponseElement,
  isResponsesElement,
  isSchemaElement,
  isBooleanJsonSchemaElement,
  isMediaTypeElement,
  isServerElement,
  isSecurityRequirementElement,
  isExternalDocumentationElement,
  isServerVariableElement,
  isContactElement,
  isExampleElement,
  isLinkElement,
  isRequestBodyElement,
  isPathsElement,
} from './predicates';

export {
  isReferenceLikeElement,
  isOpenApiExtension,
  isServerLikeElement,
  isServersElement,
} from '@swagger-api/apidom-ns-openapi-3-0';

export { keyMap, getNodeType } from './traversal/visitor';

// OpenAPI 3.1.0 elements
export {
  CallbackElement,
  ComponentsElement,
  ContactElement,
  DiscriminatorElement,
  EncodingElement,
  ExampleElement,
  ExternalDocumentationElement,
  HeaderElement,
  InfoElement,
  JsonSchemaDialectElement,
  LicenseElement,
  LinkElement,
  MediaTypeElement,
  OAuthFlowElement,
  OAuthFlowsElement,
  OpenapiElement,
  OpenApi3_1Element,
  OperationElement,
  ParameterElement,
  PathItemElement,
  PathsElement,
  ReferenceElement,
  RequestBodyElement,
  ResponseElement,
  ResponsesElement,
  SchemaElement,
  SecurityRequirementElement,
  SecuritySchemeElement,
  ServerElement,
  ServerVariableElement,
  TagElement,
  XmlElement,
} from './refractor/registration';
// NCE types
export {
  ComponentsCallbacksElement,
  ComponentsExamplesElement,
  ComponentsHeadersElement,
  ComponentsLinksElement,
  ComponentsParametersElement,
  ComponentsRequestBodiesElement,
  ComponentsResponsesElement,
  ComponentsSchemasElement,
  ComponentsSecuritySchemesElement,
  DiscriminatorMappingElement,
  EncodingHeadersElement,
  HeaderContentElement,
  HeaderExamplesElement,
  LinkParametersElement,
  MediaTypeEncodingElement,
  MediaTypeExamplesElement,
  OAuthFlowScopesElement,
  OperationCallbacksElement,
  OperationParametersElement,
  OperationSecurityElement,
  OperationServersElement,
  OperationTagsElement,
  ParameterContentElement,
  ParameterExamplesElement,
  PathItemParametersElement,
  PathItemServersElement,
  RequestBodyContentElement,
  ResponseContentElement,
  ResponseHeadersElement,
  ResponseLinksElement,
  SecurityElement,
  ServersElement,
  ServerVariablesElement,
  TagsElement,
} from '@swagger-api/apidom-ns-openapi-3-0';
export { default as ComponentsPathItemsElement } from './elements/nces/ComponentsPathItems';
export { default as WebhooksElement } from './elements/nces/Webhooks';
