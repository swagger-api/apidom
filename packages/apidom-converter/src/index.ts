import { ParseResultElement } from '@swagger-api/apidom-core';
import { mergeOptions, bundle, File } from '@swagger-api/apidom-reference';

import defaultOptions, { ConverterOptions } from './options';
import ConvertError from './errors/ConvertError';
import UnmatchedConvertStrategyError from './errors/UnmatchedConvertStrategyError';

export { ConvertError, UnmatchedConvertStrategyError };

/**
 * `convertApiDOM` already assumes that the ApiDOM is bundled.
 */
export const convertApiDOM = async (element: ParseResultElement, options = {}) => {
  const mergedOptions = mergeOptions(defaultOptions, options || {}) as ConverterOptions;
  const file = File({
    uri: mergedOptions.resolve.baseURI,
    parseResult: element,
    mediaType: mergedOptions.convert.sourceMediaType || mergedOptions.parse.mediaType,
  });
  const strategy = mergedOptions.convert.strategies.find((s) => s.canConvert(file, mergedOptions));

  if (typeof strategy === 'undefined') {
    throw new UnmatchedConvertStrategyError(file.uri);
  }

  return strategy.convert(file, mergedOptions);
};

const convert = async (uri: string, options = {}) => {
  const mergedOptions = mergeOptions(defaultOptions, options || {}) as ConverterOptions;
  const parseResult = await bundle(uri, mergedOptions);

  return convertApiDOM(parseResult, mergedOptions);
};

export default convert;
