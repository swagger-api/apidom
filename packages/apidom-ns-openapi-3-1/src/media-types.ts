import { MediaTypes } from '@swagger-api/apidom-core';

export class OpenAPIMediaTypes extends MediaTypes<string> {
  forFormat(format = 'generic') {
    const effectiveFormat = format === 'generic' ? 'openapi;version' : format;
    return this.filter((mediaType) => mediaType.includes(effectiveFormat));
  }

  latest(format = 'generic') {
    if (format === 'json') {
      return this[1];
    }
    if (format === 'yaml') {
      return this[2];
    }

    return this[0];
  }
}

const mediaTypes = new OpenAPIMediaTypes(
  'application/vnd.oai.openapi;version=3.1.0',
  'application/vnd.oai.openapi+json;version=3.1.0',
  'application/vnd.oai.openapi+yaml;version=3.1.0',
);

export default mediaTypes;
