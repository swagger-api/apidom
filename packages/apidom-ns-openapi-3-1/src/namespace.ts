import { NamespacePluginOptions } from '@swagger-api/apidom-core';
import {
  SecurityRequirementElement,
  OAuthFlowElement,
  OAuthFlowsElement,
} from '@swagger-api/apidom-ns-openapi-3-0';

import CallbackElement from './elements/Callback';
import ComponentsElement from './elements/Components';
import ContactElement from './elements/Contact';
import DiscriminatorElement from './elements/Discriminator';
import EncodingElement from './elements/Encoding';
import ExampleElement from './elements/Example';
import ExternalDocumentationElement from './elements/ExternalDocumentation';
import HeaderElement from './elements/Header';
import InfoElement from './elements/Info';
import JsonSchemaDialectElement from './elements/JsonSchemaDialect';
import LicenseElement from './elements/License';
import LinkElement from './elements/Link';
import MediaTypeElement from './elements/MediaType';
import OpenapiElement from './elements/Openapi';
import OpenApi3_1Element from './elements/OpenApi3-1';
import OperationElement from './elements/Operation';
import ParameterElement from './elements/Parameter';
import PathItemElement from './elements/PathItem';
import PathsElement from './elements/Paths';
import ReferenceElement from './elements/Reference';
import RequestBodyElement from './elements/RequestBody';
import ResponseElement from './elements/Response';
import ResponsesElement from './elements/Responses';
import SchemaElement from './elements/Schema';
import SecuritySchemeElement from './elements/SecurityScheme';
import ServerElement from './elements/Server';
import ServerVariableElement from './elements/ServerVariable';
import TagElement from './elements/Tag';
import XmlElement from './elements/Xml';

// eslint-disable-next-line @typescript-eslint/naming-convention
const openApi3_1 = {
  namespace: (options: NamespacePluginOptions) => {
    const { base } = options;

    base.register('callback', CallbackElement);
    base.register('components', ComponentsElement);
    base.register('contact', ContactElement);
    base.register('discriminator', DiscriminatorElement);
    base.register('encoding', EncodingElement);
    base.register('example', ExampleElement);
    base.register('externalDocumentation', ExternalDocumentationElement);
    base.register('header', HeaderElement);
    base.register('info', InfoElement);
    base.register('jsonSchemaDialect', JsonSchemaDialectElement);
    base.register('license', LicenseElement);
    base.register('link', LinkElement);
    base.register('mediaType', MediaTypeElement);
    base.register('oAuthFlow', OAuthFlowElement);
    base.register('oAuthFlows', OAuthFlowsElement);
    base.register('openapi', OpenapiElement);
    base.register('openApi3_1', OpenApi3_1Element);
    base.register('operation', OperationElement);
    base.register('parameter', ParameterElement);
    base.register('pathItem', PathItemElement);
    base.register('paths', PathsElement);
    base.register('reference', ReferenceElement);
    base.register('requestBody', RequestBodyElement);
    base.register('response', ResponseElement);
    base.register('responses', ResponsesElement);
    base.register('schema', SchemaElement);
    base.register('securityRequirement', SecurityRequirementElement);
    base.register('securityScheme', SecuritySchemeElement);
    base.register('server', ServerElement);
    base.register('serverVariable', ServerVariableElement);
    base.register('tag', TagElement);
    base.register('xml', XmlElement);

    return base;
  },
};

export default openApi3_1;
