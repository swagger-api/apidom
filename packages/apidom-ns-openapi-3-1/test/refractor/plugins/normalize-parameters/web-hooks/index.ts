import { expect } from 'chai';
import dedent from 'dedent';
import { sexprs } from '@swagger-api/apidom-core';
import { parse } from '@swagger-api/apidom-parser-adapter-yaml-1-2';

import { OpenApi3_1Element, refractorPluginNormalizeParameters } from '../../../../../src';

describe('refractor', function () {
  context('plugins', function () {
    context('normalize-parameters', function () {
      context('given parameters are defined in Path Item Object', function () {
        context("and Operation Object doesn't define any parameters", function () {
          specify('should inherit all Path Item parameters', async function () {
            const yamlDefinition = dedent`
              openapi: 3.1.0
              webhooks:
                hook1:
                  parameters:
                    - name: param1
                      in: query
                    - name: param2
                      in: query
                  get: {}
            `;
            const apiDOM = await parse(yamlDefinition);
            const openApiElement = OpenApi3_1Element.refract(apiDOM.result, {
              plugins: [refractorPluginNormalizeParameters()],
            });

            expect(sexprs(openApiElement)).toMatchSnapshot();
          });
        });

        context("and Operation Object doesn't defines empty parameters", function () {
          specify('should inherit all Path Item parameters', async function () {
            const yamlDefinition = dedent`
              openapi: 3.1.0
              webhooks:
                hook1:
                  parameters:
                    - name: param1
                      in: query
                    - name: param2
                      in: query
                  get:
                    parameters: []
            `;
            const apiDOM = await parse(yamlDefinition);
            const openApiElement = OpenApi3_1Element.refract(apiDOM.result, {
              plugins: [refractorPluginNormalizeParameters()],
            });

            expect(sexprs(openApiElement)).toMatchSnapshot();
          });
        });

        context('and Operation Object defines additional parameter', function () {
          specify('should merge with all Path Item parameters', async function () {
            const yamlDefinition = dedent`
              openapi: 3.1.0
              webhooks:
                hook1:
                  parameters:
                    - name: param1
                      in: query
                    - name: param2
                      in: query
                  get:
                    parameters:
                      - name: param3
                        in: query
            `;
            const apiDOM = await parse(yamlDefinition);
            const openApiElement = OpenApi3_1Element.refract(apiDOM.result, {
              plugins: [refractorPluginNormalizeParameters()],
            });

            expect(sexprs(openApiElement)).toMatchSnapshot();
          });
        });

        context('and Operation Object defines identical parameter', function () {
          specify('should replace Path Item parameter', async function () {
            const yamlDefinition = dedent`
              openapi: 3.1.0
              webhooks:
                hook1:
                  parameters:
                    - name: param1
                      in: query
                    - name: param2
                      in: query
                  get:
                    parameters:
                      - name: param1
                        in: query
                        description: operation parameter
            `;
            const apiDOM = await parse(yamlDefinition);
            const openApiElement = OpenApi3_1Element.refract(apiDOM.result, {
              plugins: [refractorPluginNormalizeParameters()],
            });

            expect(openApiElement).toMatchSnapshot();
          });
        });
      });
    });
  });
});
