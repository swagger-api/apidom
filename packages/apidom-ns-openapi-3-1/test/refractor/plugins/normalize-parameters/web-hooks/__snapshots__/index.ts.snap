// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`refractor plugins normalize-parameters given parameters are defined in Path Item Object and Operation Object defines additional parameter should merge with all Path Item parameters 1`] = `
(OpenApi3_1Element
  (MemberElement
    (StringElement)
    (OpenapiElement))
  (MemberElement
    (StringElement)
    (ObjectElement
      (MemberElement
        (StringElement)
        (PathItemElement
          (MemberElement
            (StringElement)
            (ArrayElement
              (ParameterElement
                (MemberElement
                  (StringElement)
                  (StringElement))
                (MemberElement
                  (StringElement)
                  (StringElement)))
              (ParameterElement
                (MemberElement
                  (StringElement)
                  (StringElement))
                (MemberElement
                  (StringElement)
                  (StringElement)))))
          (MemberElement
            (StringElement)
            (OperationElement
              (MemberElement
                (StringElement)
                (ArrayElement
                  (ParameterElement
                    (MemberElement
                      (StringElement)
                      (StringElement))
                    (MemberElement
                      (StringElement)
                      (StringElement)))
                  (ParameterElement
                    (MemberElement
                      (StringElement)
                      (StringElement))
                    (MemberElement
                      (StringElement)
                      (StringElement)))
                  (ParameterElement
                    (MemberElement
                      (StringElement)
                      (StringElement))
                    (MemberElement
                      (StringElement)
                      (StringElement)))))))))))
  (MemberElement
    (StringElement)
    (ObjectElement
      (MemberElement
        (StringElement)
        (ArrayElement
          (StringElement))))))
`;

exports[`refractor plugins normalize-parameters given parameters are defined in Path Item Object and Operation Object defines identical parameter should replace Path Item parameter 1`] = `
{
  "element": "openApi3_1",
  "meta": {
    "classes": {
      "element": "array",
      "content": [
        {
          "element": "string",
          "content": "api"
        },
        {
          "element": "string",
          "content": "result"
        }
      ]
    }
  },
  "content": [
    {
      "element": "member",
      "meta": {
        "classes": {
          "element": "array",
          "content": [
            {
              "element": "string",
              "content": "fixed-field"
            }
          ]
        }
      },
      "content": {
        "key": {
          "element": "string",
          "content": "openapi"
        },
        "value": {
          "element": "openapi",
          "meta": {
            "classes": {
              "element": "array",
              "content": [
                {
                  "element": "string",
                  "content": "spec-version"
                },
                {
                  "element": "string",
                  "content": "version"
                }
              ]
            }
          },
          "content": "3.1.0"
        }
      }
    },
    {
      "element": "member",
      "meta": {
        "classes": {
          "element": "array",
          "content": [
            {
              "element": "string",
              "content": "fixed-field"
            }
          ]
        }
      },
      "content": {
        "key": {
          "element": "string",
          "content": "webhooks"
        },
        "value": {
          "element": "object",
          "meta": {
            "classes": {
              "element": "array",
              "content": [
                {
                  "element": "string",
                  "content": "webhooks"
                }
              ]
            }
          },
          "content": [
            {
              "element": "member",
              "meta": {
                "classes": {
                  "element": "array",
                  "content": [
                    {
                      "element": "string",
                      "content": "patterned-field"
                    }
                  ]
                }
              },
              "content": {
                "key": {
                  "element": "string",
                  "content": "hook1"
                },
                "value": {
                  "element": "pathItem",
                  "meta": {
                    "webhook-name": {
                      "element": "string",
                      "content": "hook1"
                    }
                  },
                  "content": [
                    {
                      "element": "member",
                      "meta": {
                        "classes": {
                          "element": "array",
                          "content": [
                            {
                              "element": "string",
                              "content": "fixed-field"
                            }
                          ]
                        }
                      },
                      "content": {
                        "key": {
                          "element": "string",
                          "content": "parameters"
                        },
                        "value": {
                          "element": "array",
                          "meta": {
                            "classes": {
                              "element": "array",
                              "content": [
                                {
                                  "element": "string",
                                  "content": "path-item-parameters"
                                },
                                {
                                  "element": "string",
                                  "content": "parameters"
                                }
                              ]
                            }
                          },
                          "content": [
                            {
                              "element": "parameter",
                              "content": [
                                {
                                  "element": "member",
                                  "meta": {
                                    "classes": {
                                      "element": "array",
                                      "content": [
                                        {
                                          "element": "string",
                                          "content": "fixed-field"
                                        }
                                      ]
                                    }
                                  },
                                  "content": {
                                    "key": {
                                      "element": "string",
                                      "content": "name"
                                    },
                                    "value": {
                                      "element": "string",
                                      "content": "param1"
                                    }
                                  }
                                },
                                {
                                  "element": "member",
                                  "meta": {
                                    "classes": {
                                      "element": "array",
                                      "content": [
                                        {
                                          "element": "string",
                                          "content": "fixed-field"
                                        }
                                      ]
                                    }
                                  },
                                  "content": {
                                    "key": {
                                      "element": "string",
                                      "content": "in"
                                    },
                                    "value": {
                                      "element": "string",
                                      "content": "query"
                                    }
                                  }
                                }
                              ]
                            },
                            {
                              "element": "parameter",
                              "content": [
                                {
                                  "element": "member",
                                  "meta": {
                                    "classes": {
                                      "element": "array",
                                      "content": [
                                        {
                                          "element": "string",
                                          "content": "fixed-field"
                                        }
                                      ]
                                    }
                                  },
                                  "content": {
                                    "key": {
                                      "element": "string",
                                      "content": "name"
                                    },
                                    "value": {
                                      "element": "string",
                                      "content": "param2"
                                    }
                                  }
                                },
                                {
                                  "element": "member",
                                  "meta": {
                                    "classes": {
                                      "element": "array",
                                      "content": [
                                        {
                                          "element": "string",
                                          "content": "fixed-field"
                                        }
                                      ]
                                    }
                                  },
                                  "content": {
                                    "key": {
                                      "element": "string",
                                      "content": "in"
                                    },
                                    "value": {
                                      "element": "string",
                                      "content": "query"
                                    }
                                  }
                                }
                              ]
                            }
                          ]
                        }
                      }
                    },
                    {
                      "element": "member",
                      "meta": {
                        "classes": {
                          "element": "array",
                          "content": [
                            {
                              "element": "string",
                              "content": "fixed-field"
                            }
                          ]
                        }
                      },
                      "content": {
                        "key": {
                          "element": "string",
                          "content": "get"
                        },
                        "value": {
                          "element": "operation",
                          "meta": {
                            "http-method": {
                              "element": "string",
                              "content": "GET"
                            }
                          },
                          "content": [
                            {
                              "element": "member",
                              "meta": {
                                "classes": {
                                  "element": "array",
                                  "content": [
                                    {
                                      "element": "string",
                                      "content": "fixed-field"
                                    }
                                  ]
                                }
                              },
                              "content": {
                                "key": {
                                  "element": "string",
                                  "content": "parameters"
                                },
                                "value": {
                                  "element": "array",
                                  "meta": {
                                    "classes": {
                                      "element": "array",
                                      "content": [
                                        {
                                          "element": "string",
                                          "content": "operation-parameters"
                                        },
                                        {
                                          "element": "string",
                                          "content": "parameters"
                                        }
                                      ]
                                    }
                                  },
                                  "content": [
                                    {
                                      "element": "parameter",
                                      "content": [
                                        {
                                          "element": "member",
                                          "meta": {
                                            "classes": {
                                              "element": "array",
                                              "content": [
                                                {
                                                  "element": "string",
                                                  "content": "fixed-field"
                                                }
                                              ]
                                            }
                                          },
                                          "content": {
                                            "key": {
                                              "element": "string",
                                              "content": "name"
                                            },
                                            "value": {
                                              "element": "string",
                                              "content": "param1"
                                            }
                                          }
                                        },
                                        {
                                          "element": "member",
                                          "meta": {
                                            "classes": {
                                              "element": "array",
                                              "content": [
                                                {
                                                  "element": "string",
                                                  "content": "fixed-field"
                                                }
                                              ]
                                            }
                                          },
                                          "content": {
                                            "key": {
                                              "element": "string",
                                              "content": "in"
                                            },
                                            "value": {
                                              "element": "string",
                                              "content": "query"
                                            }
                                          }
                                        },
                                        {
                                          "element": "member",
                                          "meta": {
                                            "classes": {
                                              "element": "array",
                                              "content": [
                                                {
                                                  "element": "string",
                                                  "content": "fixed-field"
                                                }
                                              ]
                                            }
                                          },
                                          "content": {
                                            "key": {
                                              "element": "string",
                                              "content": "description"
                                            },
                                            "value": {
                                              "element": "string",
                                              "content": "operation parameter"
                                            }
                                          }
                                        }
                                      ]
                                    },
                                    {
                                      "element": "parameter",
                                      "content": [
                                        {
                                          "element": "member",
                                          "meta": {
                                            "classes": {
                                              "element": "array",
                                              "content": [
                                                {
                                                  "element": "string",
                                                  "content": "fixed-field"
                                                }
                                              ]
                                            }
                                          },
                                          "content": {
                                            "key": {
                                              "element": "string",
                                              "content": "name"
                                            },
                                            "value": {
                                              "element": "string",
                                              "content": "param2"
                                            }
                                          }
                                        },
                                        {
                                          "element": "member",
                                          "meta": {
                                            "classes": {
                                              "element": "array",
                                              "content": [
                                                {
                                                  "element": "string",
                                                  "content": "fixed-field"
                                                }
                                              ]
                                            }
                                          },
                                          "content": {
                                            "key": {
                                              "element": "string",
                                              "content": "in"
                                            },
                                            "value": {
                                              "element": "string",
                                              "content": "query"
                                            }
                                          }
                                        }
                                      ]
                                    }
                                  ]
                                }
                              }
                            }
                          ]
                        }
                      }
                    }
                  ]
                }
              }
            }
          ]
        }
      }
    },
    {
      "element": "member",
      "content": {
        "key": {
          "element": "string",
          "content": "x-normalized"
        },
        "value": {
          "element": "object",
          "content": [
            {
              "element": "member",
              "content": {
                "key": {
                  "element": "string",
                  "content": "parameters"
                },
                "value": {
                  "element": "array",
                  "content": [
                    {
                      "element": "string",
                      "content": "/webhooks/hook1/get"
                    }
                  ]
                }
              }
            }
          ]
        }
      }
    }
  ]
}
`;

exports[`refractor plugins normalize-parameters given parameters are defined in Path Item Object and Operation Object doesn't define any parameters should inherit all Path Item parameters 1`] = `
(OpenApi3_1Element
  (MemberElement
    (StringElement)
    (OpenapiElement))
  (MemberElement
    (StringElement)
    (ObjectElement
      (MemberElement
        (StringElement)
        (PathItemElement
          (MemberElement
            (StringElement)
            (ArrayElement
              (ParameterElement
                (MemberElement
                  (StringElement)
                  (StringElement))
                (MemberElement
                  (StringElement)
                  (StringElement)))
              (ParameterElement
                (MemberElement
                  (StringElement)
                  (StringElement))
                (MemberElement
                  (StringElement)
                  (StringElement)))))
          (MemberElement
            (StringElement)
            (OperationElement
              (MemberElement
                (StringElement)
                (ArrayElement
                  (ParameterElement
                    (MemberElement
                      (StringElement)
                      (StringElement))
                    (MemberElement
                      (StringElement)
                      (StringElement)))
                  (ParameterElement
                    (MemberElement
                      (StringElement)
                      (StringElement))
                    (MemberElement
                      (StringElement)
                      (StringElement)))))))))))
  (MemberElement
    (StringElement)
    (ObjectElement
      (MemberElement
        (StringElement)
        (ArrayElement
          (StringElement))))))
`;

exports[`refractor plugins normalize-parameters given parameters are defined in Path Item Object and Operation Object doesn't defines empty parameters should inherit all Path Item parameters 1`] = `
(OpenApi3_1Element
  (MemberElement
    (StringElement)
    (OpenapiElement))
  (MemberElement
    (StringElement)
    (ObjectElement
      (MemberElement
        (StringElement)
        (PathItemElement
          (MemberElement
            (StringElement)
            (ArrayElement
              (ParameterElement
                (MemberElement
                  (StringElement)
                  (StringElement))
                (MemberElement
                  (StringElement)
                  (StringElement)))
              (ParameterElement
                (MemberElement
                  (StringElement)
                  (StringElement))
                (MemberElement
                  (StringElement)
                  (StringElement)))))
          (MemberElement
            (StringElement)
            (OperationElement
              (MemberElement
                (StringElement)
                (ArrayElement
                  (ParameterElement
                    (MemberElement
                      (StringElement)
                      (StringElement))
                    (MemberElement
                      (StringElement)
                      (StringElement)))
                  (ParameterElement
                    (MemberElement
                      (StringElement)
                      (StringElement))
                    (MemberElement
                      (StringElement)
                      (StringElement)))))))))))
  (MemberElement
    (StringElement)
    (ObjectElement
      (MemberElement
        (StringElement)
        (ArrayElement
          (StringElement))))))
`;
